qir/qat/DivisionByZeroPass/DivisionByZeroPass.cpp

```diff
@@ -0,0 +1,152 @@
+
+#include <functional>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft::quantum
+{
+char* const   DivisionByZeroPass::EC_REPORT_FUNCTION      = "__qir__report_error_value";
+char* const   DivisionByZeroPass::EC_VARIABLE_NAME        = "__qir__error_code";
+int64_t const DivisionByZeroPass::EC_QIR_DIVISION_BY_ZERO = 1100;
+
+llvm::PreservedAnalyses DivisionByZeroPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& /*mam*/)
+{
+    llvm::IRBuilder<> builder(module.getContext());
+    module.getOrInsertGlobal(EC_VARIABLE_NAME, builder.getInt64Ty());
+    error_variable_ = module.getNamedGlobal(EC_VARIABLE_NAME);
+    error_variable_->setLinkage(llvm::GlobalValue::InternalLinkage);
+    error_variable_->setInitializer(builder.getInt64(0));
+    error_variable_->setConstant(false);
+
+    std::vector<llvm::Instruction*> instructions;
+    for (auto& function : module)
+    {
+        for (auto& block : function)
+        {
+            for (auto& instr : block)
+            {
+                auto* udiv = llvm::dyn_cast<llvm::SDivOperator>(&instr);
+                if (udiv)
+                {
+                    instructions.push_back(&instr);
+                }
+            }
+        }
+    }
+
+    // Injecting error code updates
+    for (auto instr : instructions)
+    {
+        auto op2 = instr->getOperand(1);
+
+        auto const& final_block = instr->getParent();
+        auto        if_block    = final_block->splitBasicBlock(instr, "if_denominator_is_zero", true);
+        auto        start_block = if_block->splitBasicBlock(if_block->getTerminator(), "-INTERMEDIATE-", true);
+        start_block->takeName(final_block);
+        final_block->setName("after_zero_check");
+
+        builder.SetInsertPoint(start_block->getTerminator());
+        auto cmp            = builder.CreateICmp(llvm::CmpInst::Predicate::ICMP_EQ, op2, builder.getInt64(0));
+        auto old_terminator = start_block->getTerminator();
+        llvm::BranchInst::Create(if_block, final_block, cmp, start_block);
+        old_terminator->eraseFromParent();
+
+        raiseError(EC_QIR_DIVISION_BY_ZERO, module, if_block->getTerminator());
+    }
+
+    // Checking error codes at end of
+    llvm::Function*                entry = nullptr;
+    std::vector<llvm::BasicBlock*> exit_blocks;
+    for (auto& function : module)
+    {
+        if (function.hasFnAttribute("EntryPoint"))
+        {
+            entry = &function;
+            for (auto& block : function)
+            {
+                auto last = block.getTerminator();
+                if (last && llvm::dyn_cast<llvm::ReturnInst>(last))
+                {
+                    exit_blocks.push_back(&block);
+                }
+            }
+            break;
+        }
+    }
+
+    if (entry)
+    {
+        for (auto start_block : exit_blocks)
+        {
+            auto if_block    = start_block->splitBasicBlock(start_block->getTerminator(), "if_error_occured", false);
```

Comment:
I assume that the GitHub warning for this line means that it should be:
```suggestion
            auto if_block    = start_block->splitBasicBlock(start_block->getTerminator(), "if_error_occurred", false);
```